newlib/libc/machine/xtensa/strcpy.c

```diff
@@ -0,0 +1,34 @@
+/* strcpy.c - Xtensa code to determine if source is PMEM or RAM and call appropriate strcpy routine.
+
+  GCC is a very smart compiler, and it will, in fact, replace printf(), and its
+  related functions with strcpy() calls in order to optimize speed.  This fails
+  horribly when printf(PSTR("xxx")) is called since the original strcpy can't
+  handle the PROGMEM source.
+
+  See http://www.ciselant.de/projects/gcc_printf/gcc_printf.html for more info.
+
+  There are two ways around this:
+  - Adding -fno-builtin-*printf, which slows down every printf() call that was
+    being optimized before.  GCC won't replace the printf() call with a strcpy()
+    call, and everything will work since we now support PROGMEM strings in
+    printf().
+  - Make strcpy smarter and fall back on the appropriate routine depending on
+    the source parameter.  Since on the ESP8266 PROGMEM starts at 0x40000000
+    this is a simple comparison.  In this case speed will be maintained.
+*/
+
+#include <string.h>
+
+extern char *__fast_strcpy(char *dest, const char *src);
+char *strcpy(char *dest, const char *src)
+{
+    if (src >= (const char *)0x40000000) return strcpy_P(dest, src);
```

Comment:
This checks the lower mem mapped boundary. Should an upper boundary be checked as well?

Reply:
IMHO it's not needed.  I'm not really ecstatic about this comparison, BTW, but GCC's a sneaky one.  Took me a while to figure out why only some printf(PSTR(),...) were failing in strcpy (which isn't called by vfprintf)!

Granted the docs are pretty poor, from what I can see all RAM on the ESP2866 lives below 4M.  If you go above progmem I imagine you'd get a LOAD ERROR anyway, and if not then it should always be safe to do 32-bit-only reads from any memory space.  On the ESP32 w/the "MMU windows" this may be different, but I've not looked deeply at how it's handled there.

If you're doing a strcpy from register space, well, you've got bigger problems than this! :)

Reply:
I agree with the test. It is sufficient to check whether a valid pointer is in ram or flash.

Reply:
Just for the future reference, the CPU inside ESP8266 has several ports, listed here:
https://github.com/igrr/newlib-xtensa/blob/e6743695d217592a298c51ecb410cfab1dcdc167/newlib/libc/sys/xtensa/include/xtensa/config/core-isa.h#L213-L253

(names are a bit confusing, for example the 0x40000000 — 0x40100000 regions is called "Instruction RAM 0" but it is actually connected to the mask ROM)

By default (after CPU reset) you don't get a load error when trying to access memory outside of these regions, but CPU's region protection feature can be configured to disallow access to certain (256MB sized) regions. I think the SDK does this by default as part of the startup code. I can't link to this code for obvious reasons, but here's the equivalent for the ESP32: 
https://github.com/espressif/esp-idf/blob/3bc1a8318b95d13f2b3ecdd3fb0545b1471863ba/components/soc/esp32/include/soc/cpu.h#L54-L74

Reply:
The docs I've read are a bit light, is there some listing of what types of accesses are allowed?  IROMs/IRAMS seem to indicate they require 32-bit only accesses (as they're supposed to store instructions...).  Replacing 0x400000 with min(IRAM0VADDR, IRAM0VADDR, IRAM1VADDR) is where you're going with this?